libs/sync/snapshot_state.go

```diff
@@ -0,0 +1,115 @@
+package sync
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/databricks/cli/libs/fileset"
+	"github.com/databricks/cli/libs/notebook"
+)
+
+// SnapshotState keeps track of files on the local filesystem and their corresponding
+// entries in WSFS.
+type SnapshotState struct {
+	// Map of local file names to their last recorded modified time. Files found
+	// to have a newer mtime have their content synced to their remote version.
+	LastModifiedTimes map[string]time.Time `json:"last_modified_times"`
+
+	// Map of local file names to corresponding remote names.
+	// For example: A notebook named "foo.py" locally would be stored as "foo"
+	// in WSFS, and the entry would be: {"foo.py": "foo"}
+	LocalToRemoteNames map[string]string `json:"local_to_remote_names"`
+
+	// Inverse of LocalToRemoteNames. Together they form a 1:1 mapping where all
+	// the remote names and local names are unique.
+	RemoteToLocalNames map[string]string `json:"remote_to_local_names"`
+}
+
+// Convert an array of files on the local file system to a SnapshotState representation.
+func NewSnapshotState(localFiles []fileset.File) (*SnapshotState, error) {
+	fs := &SnapshotState{
+		LastModifiedTimes:  make(map[string]time.Time),
+		LocalToRemoteNames: make(map[string]string),
+		RemoteToLocalNames: make(map[string]string),
+	}
+
+	// Expect no files to have a duplicate entry in the input array.
+	seen := make(map[string]fileset.File)
+	for _, f := range localFiles {
+		if _, ok := seen[f.Relative]; !ok {
+			seen[f.Relative] = f
+		} else {
+			return nil, fmt.Errorf("expected only one entry per file. Found duplicate entries for file: %s", f.Relative)
+		}
+	}
+
+	// Compute the new state.
+	for _, f := range localFiles {
+		// Compute the remote name the file will have in WSFS
+		remoteName := filepath.ToSlash(f.Relative)
+		isNotebook, _, err := notebook.Detect(f.Absolute)
+		if err != nil {
+			// TODO: Add test that notebook ignore works as expected.
+			// Ignore this file if we're unable to determine the notebook type.
+			// Trying to upload such a file to the workspace would fail anyway.
+			return nil, nil
+		}
+		if isNotebook {
+			ext := filepath.Ext(remoteName)
+			remoteName = strings.TrimSuffix(remoteName, ext)
+		}
+
+		// Add the file to snapshot state
+		fs.LastModifiedTimes[f.Relative] = f.Modified()
+		if existingLocalName, ok := fs.RemoteToLocalNames[remoteName]; ok {
+			return nil, fmt.Errorf("both %s and %s point to the same remote file location %s. Please remove one of them from your local project", existingLocalName, f.Relative, remoteName)
+		}
+
+		fs.LocalToRemoteNames[f.Relative] = remoteName
+		fs.RemoteToLocalNames[remoteName] = f.Relative
+	}
+	return fs, fs.validate()
```

Comment:
What's the value of validating immediately after construction?

Reply:
The validate function makes sure the invariants we expect out of the state hold true. The infinite sync issue is a good example of a case where our these invariants were voilated and downstream code had wierd behaviour.

This makes sure we fail early. I think there's value in this approach because if there's an edge case we miss, this will complain loudly. This also allows downstream code to be confident about certain properties in the state, like `local_to_remote` and `remote_to_local` maps form a bijective function.

Reply:
@shreyas-goenka if you make `validate` just a unit test you will have the same, you won't be able to make a breaking change anyway and won't need to do the check in runtime. But with unit test you catch the error earlier comparing to runtime validate check. 

The runtime invariant checks make sense if we construct the data based on input from files or users, is it something that is planned?